### DISCUSSION TITLE
How to use LUA to get input name as "P1_JOYSTICK_UP" ?

### DISCUSSION THREAD

**Discussion by Jakobud** [mamedev/discussions#147] [Q&A] [answered]
I'm messing around with the LUA scripting, trying to figure out a way to dump out all the input assignments. I'm looking at the ioport_field type. I'm trying this:

```
		for pname, port in pairs(manager.machine.ioport.ports) do

			for fname, field in pairs(port.fields) do
				print(field.name)
			end
		end
```

I'm trying to get values like "P1_BUTTON1", "P2_JOYSTICK_LEFT", etc. But `field.name` returns things like "P1 UP" or "High Punch" (mortal kombat). I have been looking at the field table properties but don't see what I am looking for.

What part of the LUA API do I need to look at to get this information? Thanks,

**Comment by sairuk** [comment]
They are the internal port names MAME uses, I believe you'll need to provide your own mapping atm if you want to use other naming. Although i'd also be interested to hear otherwise. 

It all starts to get interesting when you start to look at fields for keyboards as I have started to do [here](https://github.com/mameau/mame-plugins/tree/main/plugins/act486auto/portmap)

**Reply by cuavas** [reply]
Nah, they’re input type tokens – the I/O port manager has methods for dealing with them: https://docs.mamedev.org/luascript/ref-input.html#i-o-port-manager

They’re used by various plugins to support input configuration.

**Reply by sairuk** [reply]
Thanks, i'll have to take a closer look into input type tokens

**Comment by cuavas** [comment, marked as answer]
`field.name` is the display name, you want an input type token.  Remember they aren’t necessarily unique within a device.

```lua
mach = manager.machine
portman = mach.ioport
for tag, port in pairs(portman.ports) do
    for name, field in pairs(port.fields) do
        print(portman:input_type_to_token(field.type, field.player))
    end
end
```

**Reply by Jakobud** [reply]
Ah that is exactly what I was looking for thank you. I wish the LUA docs were a little more organized and had examples for every class and property.

**Reply by cuavas** [reply]
Examples wouldn't help if you're looking at the wrong property.

**Reply by Jakobud** [reply]
Well I would have figured it out eventually just by searching around in the docs :-) It just wasn't clear to me what the difference was between say a input_type and a token for example.

Is the docs.mamedev.org in a repo that I could look at and help contribute to it? I don't see it in the www.mamedev.org repo unless I'm just not looking hard enough.

Thanks for the help

**Reply by Jakobud** [reply]
One more related question, is it possible to query any of the above information without actually loading a rom? Like I see you can use `emu.driver_find(name)` to look up metadata for a game, but I don't see anything there about input information. I figure that data is in the drivers whether or not you actually have the rom for a game, so perhaps there is a way to access it using the LUA api... Any thoughts?

Thanks again!

**Reply by cuavas** [reply]
No, it isn't possible. There are a bunch of factors that mean you can't really get that information without a complete machine configuration.
